include/deal.II/lac/sparse_matrix.templates.h

```diff
@@ -1008,37 +1008,10 @@ SparseMatrix<number>::mmult (SparseMatrix<numberC>       &C,
       C.clear();
       sp_C.reinit (0,0,0);
 
-      // create a sparsity pattern for the matrix. we will go through all the
-      // rows in the matrix A, and for each column in a row we add the whole
-      // row of matrix B with that row number. This means that we will insert
-      // a lot of entries to each row, which is best handled by the
-      // DynamicSparsityPattern class.
+      // create a sparsity pattern for the matrix C.
       {
-        DynamicSparsityPattern dsp (m(), B.n());
-        for (size_type i = 0; i < dsp.n_rows(); ++i)
-          {
-            const size_type *rows = &sp_A.colnums[sp_A.rowstart[i]];
-            const size_type *const end_rows =
-              &sp_A.colnums[sp_A.rowstart[i+1]];
-            for (; rows != end_rows; ++rows)
-              {
-                const size_type col = *rows;
-                size_type *new_cols = const_cast<size_type *>
-                                      (&sp_B.colnums[sp_B.rowstart[col]]);
-                size_type *end_new_cols = const_cast<size_type *>
-                                          (&sp_B.colnums[sp_B.rowstart[col+1]]);
-
-                // if B has a diagonal, need to add that manually. this way,
-                // we maintain sortedness.
-                if (sp_B.n_rows() == sp_B.n_cols())
```

Comment:
i don't know why this line is needed, but the new function you added https://github.com/dealii/dealii/blob/master/source/lac/dynamic_sparsity_pattern.cc#L422-L437 

it's probably related to `indices_are_unique_and_sorted=true` in the `add_entries`. 

I am sure others know better, but the new code is not one-to-one the same, but might be indeed equivalent.

Reply:
I think the new code is equivalent, it just passes through the data in a different order. This one here tried to be clever and accesses the diagonal separately, but that is not needed actually.